src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -34,19 +36,66 @@ public HostingDiagnosticListener(TelemetryClient client, ICorrelationIdLookupHel
+        public void OnHttpRequestIn()
+        {
+            // do nothing, just enable the diagnotic source
+        }
+
+        /// <summary>
+        /// Diagnostic event handler method for 'Microsoft.AspNetCore.Hosting.HttpRequestIn.Start' event.
+        /// </summary>
+        [DiagnosticName("Microsoft.AspNetCore.Hosting.HttpRequestIn.Start")]
+        public void OnHttpRequestInStart(HttpContext httpContext)
         {
             if (this.client.IsEnabled())
             {
-                var requestTelemetry = new RequestTelemetry
+                var requestTelemetry = new RequestTelemetry();
+
+                if (Activity.Current == null)
                 {
-                    Id = httpContext.TraceIdentifier
-                };
+                    AspNetCoreEventSource.Instance.LogHostingDiagnosticListenerOnHttpRequestInStartActivityNull();
+                }
+                else if (Activity.Current.ParentId != null)
+                {
+                    // The activity is initialized from the new "Request-Id" header
+                    requestTelemetry.Context.Operation.ParentId = Activity.Current.ParentId;
+
+                    foreach (var prop in Activity.Current.Baggage)
+                    {
+                        if (!requestTelemetry.Context.Properties.ContainsKey(prop.Key))
+                        {
+                            requestTelemetry.Context.Properties[prop.Key] = prop.Value;
+                        }
+                    }
+                }
+                else
+                {
+                    // The request doesn't have the new "Request-Id" header but the standard headers,
+                    // we create a new activity and initialize it by the standard headers.
+                    var xmsRequestRootId = httpContext?.Request?.Headers[RequestResponseHeaders.StandardRootIdHeader];
+                    if (xmsRequestRootId.HasValue && xmsRequestRootId.Value.Count > 0)
+                    {
+                        var activity = new Activity(ActivityInitializedByStandardHeaderName);
+                        activity.SetParentId(xmsRequestRootId.Value[0]);
+                        activity.Start();
+                        httpContext.Features.Set(activity);
+                    }
+
+                    var xmsRequestId = httpContext?.Request?.Headers[RequestResponseHeaders.StandardParentIdHeader];
+                    if (xmsRequestId.HasValue && xmsRequestId.Value.Count > 0)
+                    {
+                        requestTelemetry.Context.Operation.ParentId = xmsRequestId.Value[0];
+                    }
+                }
+
+                requestTelemetry.Id = Activity.Current?.Id;
+                requestTelemetry.Context.Operation.Id = Activity.Current?.RootId;
+
                 this.client.Initialize(requestTelemetry);
-                requestTelemetry.Start(timestamp);
+                requestTelemetry.Start(Stopwatch.GetTimestamp());
```

Comment:
They don't pass the timestamp because the created Activity has StartTimeUtcNow property. But you're right, it feels less accurate. I can create an issue for AspNetCore if you think it's important.